ewms_pilot/pilot.py

```diff
@@ -311,34 +308,102 @@ async def consume_and_reply(
         raise
 
 
-async def _ack_nack_finished_tasks(
+async def _wait_on_tasks_with_ack(
     sub: mq.queue.ManualQueueSubResource,
-    tasks: AsyncioTaskMessages,
-    return_when: str,
-    previous_failed: AsyncioTaskMessages,
-) -> Tuple[AsyncioTaskMessages, AsyncioTaskMessages]:
+    pub: mq.queue.QueuePubResource,
+    tasks_msgs: AsyncioTaskMessages,
+    return_when_all_done: bool,
+    previous_task_errors: List[BaseException],
+    # TODO: replace when https://github.com/Observation-Management-Service/MQClient/issues/56
+    rabbitmq_raw_queues: Optional[List["mq.broker_clients.rabbitmq.RabbitMQ"]] = None,
+) -> Tuple[AsyncioTaskMessages, List[BaseException]]:
     """Get finished tasks and ack/nack their messages.
 
     Returns:
         Tuple:
             AsyncioTaskMessages: pending tasks and
-            AsyncioTaskMessages: failed tasks (plus those in `previous_failed`)
+            List[BaseException]: failed tasks' exceptions (plus those in `previous_task_errors`)
     """
-    done, pending = await asyncio.wait(tasks.keys(), return_when=return_when)
-    LOGGER.info(f"{len(done)} Tasks Finished")
-
-    for task in done:
-        if task.exception():
-            await sub.nack(tasks[task])
-            previous_failed[task] = tasks[task]
-            LOGGER.error("Task failed:")
-            LOGGER.error(_task_exception_str(task))
-        else:
-            await sub.ack(tasks[task])
+    pending: Set[asyncio.Task] = set(tasks_msgs.keys())  # type: ignore[type-arg]
+
+    async def handle_failed_task(task: asyncio.Task, exception: BaseException) -> None:  # type: ignore[type-arg]
+        previous_task_errors.append(exception)
+        LOGGER.error(
+            f"TASK FAILED ({repr(exception)}) -- attempting to nack original message..."
+        )
+        try:
+            await sub.nack(tasks_msgs[task])
+        except Exception as e:
+            # LOGGER.exception(e)
+            LOGGER.error(f"Could not nack: {repr(e)}")
+        LOGGER.error(_all_task_errors_string(previous_task_errors))
+
+    # LOOP!
+    while pending:
+        # looping over asyncio.FIRST_COMPLETED is like asyncio.ALL_COMPLETED
+
+        # alert rabbitmq  # TODO: replace when https://github.com/Observation-Management-Service/MQClient/issues/56
+        if rabbitmq_raw_queues:
+            for raw_q in rabbitmq_raw_queues:
+                if raw_q.connection:
+                    LOGGER.info("sending heartbeat to RabbitMQ broker...")
+                    raw_q.connection.process_data_events()
+
+        # wait for next task
+        done, pending = await asyncio.wait(
+            pending,
+            return_when=asyncio.FIRST_COMPLETED,
+            timeout=_HOUSEKEEPING_TIMEOUT,
+        )
+
+        # HANDLE FINISHED TASK(S)
+        for task in done:  # fyi, should just be one task in here max
+            # FAILED TASK
+            if e := task.exception():
```

Comment:
It feels slightly weird to me to need `.exception()` and `.result()`, instead of awaiting the task and catching the error.  But I guess if there's a reason, it's fine.

Reply:
There wasn't a reason -- I've updated that block and removed some nesting too